### PR TITLE
Add support to schedule Timer and Appointment in daemon mode

### PR DESCRIPTION
**Pull request by Shan1024**
## Purpose
This PR updates timer and appointment to be scheduled in daemon mode. Caching package is updated so that the cache clearing task is run in daemon mode.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal [hunk 1]
```diff
@@ -20,19 +20,20 @@ package ballerina.task;
 @Param {value:"onTrigger: The function which gets called when the timer goes off"}
 @Param {value:"onError: The function that gets called if the onTrigger function returns an error"}
 @Param {value:"schedule: Specifies the initial delay and interval of the timer task"}
+@Param {value:"isDaemon: Specifies whether the timer should run in daemon mode"}
 @Return {value:"The unique ID of the timer task that was scheduled"}
-public native function scheduleTimer ((function() returns error?) onTrigger,
-                                      (function(error) returns ())? onError,
-                                      {int delay = 0; int interval;} schedule) returns string;
+public native function scheduleTimer ((function() returns error?) onTrigger, (function(error) returns ())? onError,
+                                      {int delay = 0; int interval;} schedule, boolean isDaemon) returns string;
```
Comment (afkham) [inline stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: I am not very comfortable with this. For an external user, isDaemon is very confusing.

Comment (Shan1024) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: Shall we change the parameter name then?

Comment (afkham) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: This should be a defaultable param. Also, does the user have to bother about this? Can we always make it a daemon?

Comment (Shan1024) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: @lafernando - Can we run these in daemon mode only? WDYT?

Comment (lafernando) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: No .. that is the problem we had originally .. earlier it was effectively running in daemon mode. That is, the timer will get destroyed and the process with it, when the main thread exits. But for timers, we need to generally work even when the main thread exits. This is the case, when users registers timers with a callback, and then they work. But for scenarios as Ballerina Cache and Transactions, it should work in the other way, where as, when the main thread exists, the internal timers used in Cache etc.. should get destroyed, and the process should exit. 

So due to this, the default behavior of timers should be non-daemon, where people can directly use the Timer API, and for other Ballerina features which depend on it, such as Cache etc.. they can pass in the isDaemon property, to make it true, to make it work in the way they want.

Comment (afkham) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: Let's hold this a bit. I want to discuss and understand further before making this API change.

**Comment by Shan1024** [comment]
Closing since we haven't finalized discussions on daemon mode yet.

**Comment by lafernando** [comment]
We need to add this. Shall we name that property something else, if the name is problematic. How about "killOnExit"?. Where, if killOnExit would be false by default, but in places like Cache, we can make it true.

**Comment by afkham** [comment]
@lafernando @Shan1024 @sanjiva this looks to me more like an implementation detail. The user will not see any value in setting this variable and will have to understand how the process is going to run in order to decide what to set. When we schedule tasks in Java, for example, we don't have to do anything like this. So, this behavior should not be exposed in the task API.

**Comment by lafernando** [comment]
@afkham @sanjiva Actually as it seems, Java also have the same property in their timers (https://docs.oracle.com/javase/7/docs/api/java/util/Timer.html#Timer(boolean)). 

Either way, the requirement we have is, since we use the same timer API, in Cache and all (as we should) we need to have this behavior somehow in our timer API. And it has to be a property that needs to be exposed to the outside, and not an implementation detail of the timer API, since it is used by other dependent Ballerina libraries/applications. 

If we do things like, doing a sleep or something in our main application to prolong the execution, so the timer can keep executing, will look like a not-so-nice hack. So we need to implement this properly.

File: stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal (hunk 1 above)
Comment (sanjiva) [reply stdlib/ballerina-builtin/src/main/ballerina/task/natives.bal]: Yes +1 for reviewing .. daemon is a Java concept and we should find an alternative that works for us.

**Comment by uthaiyashankar** [comment]
Closing this based on offline chats. This needs design review, resolving conflicts, etc.

**Pull request closed without merging**